electricitymap/contrib/libs/models/datapoints.py

```diff
@@ -0,0 +1,427 @@
+        datetime: datetime,
+        source: str,
+        production: ProductionMix,
+        storage: Optional[StorageMix] = None,
+        forecasted: bool = False,
+    ) -> Optional["ProductionBreakdown"]:
+        try:
+            # Correct negative production values.
+            for key, value in production.dict().items():
+                if value is not None and value < 0:
+                    production.__setattr__(key, None)
+                    logger.warning(
+                        f"Production value for {key} is negative.\
+                              This value is set to None."
+                    )
+            return ProductionBreakdown(
+                zoneKey=zoneKey,
+                datetime=datetime,
+                source=source,
+                production=production,
+                storage=storage,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating production breakdown Event {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "production": self.production.dict(exclude_none=True),
+            "storage": self.storage.dict(exclude_none=True) if self.storage else None,
+            "source": self.source,
+        }
+
+
+class Consumption(Event):
+    consumption: float
+
+    @validator("consumption")
+    def _validate_consumption(cls, v: float):
+        if v < 0:
+            raise ValueError(f"Consumption cannot be negative: {v}")
+        if v > 500000:
+            raise ValueError(f"Consumption is implausibly high, above 500GW: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zoneKey: ZoneKey,
+        datetime: datetime,
+        source: str,
+        consumption: float,
+        forecasted: bool = False,
+    ) -> Optional["Consumption"]:
+        try:
+            return Consumption(
+                zoneKey=zoneKey,
+                datetime=datetime,
+                source=source,
+                consumption=consumption,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(
+                f"Error creating consumption Event {datetime}: {e}",
+                extra={
+                    "zoneKey": zoneKey,
+                    "datetime": datetime,
+                    "kind": "consumption",
+                },
+            )
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "consumption": self.consumption,
+            "source": self.source,
+        }
+
+
+class Price(Event):
+    price: float
+    currency: str
+
+    @validator("currency")
+    def _validate_currency(cls, v: str):
+        if v not in VALID_CURRENCIES:
+            raise ValueError(f"Unknown currency: {v}")
+        return v
```

Comment:
Right now this is a bit strict, I think we should allow lowercase matches as well and convert them to uppercase in here so it's consistent in the backend.

Reply:
Since we are going to spend time updating every parser, I think it can be worth it to standardize and be strict about it.